lib/os_io.c

```diff
@@ -37,6 +37,197 @@ ___io_module ___io_mod =
 
 /*---------------------------------------------------------------------------*/
 
+/* poll dynamic fdset memory management. */
+
+#ifdef USE_select_or_poll
+
+___HIDDEN int ___fdset_init
+   ___P((___processor_state ps),
+        (ps)
+        ___processor_state ps;)
+{
+  void *readfds = NULL;
+  void *writefds = NULL;
+  void *exceptfds = NULL;
+
+  int size = ___VMSTATE_FROM_PSTATE(ps)->os.fdset.size;
+  int bytes = ___CEILING_DIV (size, 8);
+
+  readfds  = ___ALLOC_MEM (bytes);
+  if (readfds == NULL)
+    goto error;
+
+  writefds = ___ALLOC_MEM (bytes);
+  if (writefds == NULL)
+    goto error;
+
+#ifdef USE_select
+  exceptfds = ___ALLOC_MEM (bytes);
+  if (exceptfds == NULL)
+    goto error;
+#endif
+
+  ps->os.fdset.readfds = readfds;
+  ps->os.fdset.writefds = writefds;
+  ps->os.fdset.exceptfds = exceptfds;
+  ps->os.fdset.size = size;
+
+  return 1;
+
+ error:
+  if (readfds != NULL)
+    ___FREE_MEM (readfds);
+  if (writefds != NULL)
+    ___FREE_MEM (writefds);
+  if (exceptfds != NULL)
+    ___FREE_MEM (exceptfds);
+
+  return 0;
+}
```

Comment:
It would be a better style for ```___fdset_init``` to call ```___fdset_realloc``` after initializing fields to ```NULLL``` to avoid repeated logic.